File: src/utils/imageProcessing.ts
import sharp from "sharp";
import { RasterImage } from "../types";
import { getPrintableWidthPixels, getPaperWidthMm, mmToPixels } from "./printerSettings";

interface PrepareImageParams {
  filePath: string;
  paper: 58 | 80;
  dpi: number;
  threshold: number;
}

function packMonoBitmap(gray: Buffer, width: number, height: number, threshold: number): Buffer {
  const bytesPerRow = width / 8;
  const output = Buffer.alloc(bytesPerRow * height);

  for (let y = 0; y < height; y += 1) {
    for (let xByte = 0; xByte < bytesPerRow; xByte += 1) {
      let byte = 0;

      for (let bit = 0; bit < 8; bit += 1) {
        const x = xByte * 8 + bit;
        const pixel = gray[y * width + x];
        const isBlack = pixel < threshold;
        if (isBlack) {
          byte |= 0x80 >> bit;
        }
      }

      output[y * bytesPerRow + xByte] = byte;
    }
  }

  return output;
}

export async function prepareRasterImage(params: PrepareImageParams): Promise<RasterImage> {
  const targetWidth = getPrintableWidthPixels(params.paper, params.dpi);
  const image = sharp(params.filePath).rotate();
  const metadata = await image.metadata();

  if (!metadata.width || !metadata.height) {
    throw new Error("File gambar tidak valid.");
  }

  const resizedHeight = Math.max(1, Math.round((metadata.height * targetWidth) / metadata.width));
  const { data } = await image
    .resize({ width: targetWidth, height: resizedHeight, fit: "inside" })
    .grayscale()
    .raw()
    .toBuffer({ resolveWithObject: true });

  const bytesPerRow = targetWidth / 8;
  const bitmap = packMonoBitmap(data, targetWidth, resizedHeight, params.threshold);

  return {
    width: targetWidth,
    height: resizedHeight,
    bytesPerRow,
    bitmap
  };
}

export async function preparePrintablePng(
  filePath: string,
  paper: 58 | 80,
  dpi: number
): Promise<{ pngBuffer: Buffer; widthMm: number; heightMm: number }> {
  const targetWidth = getPrintableWidthPixels(paper, dpi);
  const image = sharp(filePath).rotate();
  const metadata = await image.metadata();

  if (!metadata.width || !metadata.height) {
    throw new Error("File gambar tidak valid.");
  }

  const resizedHeight = Math.max(1, Math.round((metadata.height * targetWidth) / metadata.width));
  const pngBuffer = await image.resize({ width: targetWidth, height: resizedHeight, fit: "inside" }).png().toBuffer();

  const widthMm = getPaperWidthMm(paper);
  const heightMm = (resizedHeight / dpi) * 25.4;

  // Fallback in case metadata or resize behaves unexpectedly.
  if (!Number.isFinite(heightMm) || heightMm <= 0) {
    return {
      pngBuffer,
      widthMm,
      heightMm: (mmToPixels(widthMm, dpi) / dpi) * 25.4
    };
  }

  return { pngBuffer, widthMm, heightMm };
}

File: printer_tool/image_processing.py
from dataclasses import dataclass

from PIL import Image, ImageOps
from PIL import ImageEnhance

from .settings import get_paper_width_mm, get_printable_width_pixels


@dataclass
class RasterImage:
    width: int
    height: int
    bytes_per_row: int
    bitmap: bytes


def _pack_mono_bitmap(gray_image: Image.Image, threshold: int) -> bytes:
    width, height = gray_image.size
    pixels = gray_image.tobytes()
    bytes_per_row = width // 8
    out = bytearray(bytes_per_row * height)

    for y in range(height):
        for x_byte in range(bytes_per_row):
            value = 0
            for bit in range(8):
                x = x_byte * 8 + bit
                pixel = pixels[y * width + x]
                if pixel < threshold:
                    value |= 0x80 >> bit
            out[y * bytes_per_row + x_byte] = value

    return bytes(out)


def _apply_darkness(image: Image.Image, darkness: int) -> Image.Image:
    if darkness < 50 or darkness > 180:
        raise ValueError("darkness harus 50-180.")

    # 100 = netral, >100 lebih gelap, <100 lebih terang.
    brightness_factor = 100.0 / float(darkness)
    enhancer = ImageEnhance.Brightness(image)
    return enhancer.enhance(brightness_factor)


def prepare_raster_image(file_path: str, paper: int, dpi: int, threshold: int, darkness: int = 100) -> RasterImage:
    if threshold < 0 or threshold > 255:
        raise ValueError("threshold harus 0-255.")

    target_width = get_printable_width_pixels(paper, dpi)
    image = Image.open(file_path)
    image = ImageOps.exif_transpose(image)
    image = image.convert("L")

    ratio = image.height / image.width
    resized_height = max(1, round(target_width * ratio))
    image = image.resize((target_width, resized_height), Image.Resampling.LANCZOS)
    image = _apply_darkness(image, darkness)

    bitmap = _pack_mono_bitmap(image, threshold)
    return RasterImage(
        width=target_width,
        height=resized_height,
        bytes_per_row=target_width // 8,
        bitmap=bitmap,
    )


def prepare_printable_rgb(file_path: str, paper: int, dpi: int, darkness: int = 100) -> tuple[Image.Image, float, float]:
    target_width = get_printable_width_pixels(paper, dpi)
    width_mm = float(get_paper_width_mm(paper))

    image = Image.open(file_path)
    image = ImageOps.exif_transpose(image)
    image = image.convert("RGB")

    ratio = image.height / image.width
    resized_height = max(1, round(target_width * ratio))
    image = image.resize((target_width, resized_height), Image.Resampling.LANCZOS)
    image = _apply_darkness(image, darkness)
    height_mm = (resized_height / dpi) * 25.4

    return image, width_mm, height_mm
